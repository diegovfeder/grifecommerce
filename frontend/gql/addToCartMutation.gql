	mutation AddToCartMutation($id: ID!) {
		addToCart(productId: $id) {
			id
		}
	}
