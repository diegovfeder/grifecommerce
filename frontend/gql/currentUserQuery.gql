query CurrentUser {
		authenticatedItem {
			... on User {
				id
				email
				name
				cart {
					id
					quantity
					product {
						id
						price
						name
						description
						photo {
							image {
								publicUrlTransformed
							}
						}
					}
				}
			}
		}
	}
