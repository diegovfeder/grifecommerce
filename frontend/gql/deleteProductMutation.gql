mutation DELETE_PRODUCT_MUTATION($id: ID!) {
	deleteProduct(id: $id) {
		id
		name
	}
}
