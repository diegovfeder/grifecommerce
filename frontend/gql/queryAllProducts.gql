query AllProducts {
	products {
		id
		name
		price
		description
		photo {
			id
			image {
				publicUrlTransformed
			}
		}
	}
}
