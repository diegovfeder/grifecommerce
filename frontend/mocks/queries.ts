import { CURRENT_USER_QUERY } from '../hooks/useUserQuery';

const mockQueries = {
	currentUser: {
		request: {
			query: CURRENT_USER_QUERY,
			variables: {},
		},
		result: {
			data: {
				authenticatedItem: {
					id: '1',
					email: 'diego@test.com',
					name: 'Diego Test',
					cart: {
						id: '1',
						quantity: '1',
						product: {
							id: '1',
							price: '200',
							name: 'Sample Pack #01',
							description: 'This is a sample pack',
							photo: {
								image: {
									publicUrlTransformed: 'https://via.placeholder.com/300x300',
								},
							},
						},
					},
				},
			},
		},
	},
};

export default mockQueries;
