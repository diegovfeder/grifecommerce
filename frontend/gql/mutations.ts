import gql from 'graphql-tag';

export const SEND_USER_PASSWORD_RESET_LINK_MUTATION = gql`
	mutation SendUserPasswordResetLink($email: String!) {
		sendUserPasswordResetLink(email: $email)
	}
`;

export const SIGN_UP_MUTATION = gql`
	mutation SIGNUP_MUTATION(
		$email: String!
		$name: String!
		$password: String!
	) {
		createUser(data: { email: $email, name: $name, password: $password }) {
			id
			email
			name
		}
	}
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CreateOrder($token: String!) {
		checkout(token: $token) {
			id
			charge
			total
			items {
				id
				name
			}
		}
	}
`;
