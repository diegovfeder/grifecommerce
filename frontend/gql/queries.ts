import gql from 'graphql-tag';

export const USER_EMAIL_QUERY = gql`
	query UserEmail($email: String!) {
		user(where: { email: $email }) {
			id
			email
		}
	}
`;

export const PRODUCT_QUERY = gql`
	query Product($id: ID!) {
		product(where: { id: $id }) {
			id
			name
			price
			description
			photo {
				id
				altText
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const ALL_PRODUCTS_QUERY = gql`
	query AllProducts {
		products {
			id
			name
			price
			description
			photo {
				id
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const PRODUCTS_QUERY = gql`
	query Products($take: Int, $skip: Int = 0) {
		products(take: $take, skip: $skip) {
			id
			name
			price
			description
			photo {
				id
				image {
					publicUrlTransformed
				}
			}
		}
	}
`;

export const PRODUCTS_COUNT_QUERY = gql`
	query ProductsCount {
		productsCount
	}
`;

export const CURRENT_USER_QUERY = gql`
	query CurrentUser {
		authenticatedItem {
			... on User {
				id
				email
				name
				cart {
					id
					quantity
					product {
						id
						price
						name
						description
						photo {
							image {
								publicUrlTransformed
							}
						}
					}
				}
			}
		}
	}
`;
