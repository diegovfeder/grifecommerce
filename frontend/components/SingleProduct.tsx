import { useQuery } from '@apollo/client';
import styled from 'styled-components';
import Image from 'next/image';
import Head from 'next/head';
import ErrorMessage from './ErrorMessage';
import { PRODUCT_QUERY } from '../gql/queries';
import { LoadingSkeleton } from './loading';
import Supreme, { SupremeDescription } from './styles/Supreme';
import formatMoney from '../utils/formatMoney';
import { TEXT_NO_DESCRIPTION } from '../utils/constants';

interface SingleProductProps {
	id: string;
}

const SingleProduct = ({ id }: SingleProductProps) => {
	const { loading, data, error } = useQuery(PRODUCT_QUERY, {
		variables: {
			id,
		},
		onCompleted: data => {
			console.log({ data });
		},
	});

	if (loading) return <LoadingSkeleton />;

	if (error) return <ErrorMessage error={error} />;

	const { product } = data;
	const isPhotoImageUrlDefined = !!product.photo?.image?.publicUrlTransformed;

	return (
		<ProductStyles data-test-id="singleProduct">
			<Head>
				<title>GRIFE | {product.name}</title>
			</Head>
			<div style={{ display: 'flex', flex: 1, flexDirection: 'column' }}>
				<h1 style={{ borderBottom: '4px solid gray', fontWeight: '100' }}>
					{product.name}
				</h1>
				<Supreme>{formatMoney(product.price)}</Supreme>
				{isPhotoImageUrlDefined ? (
					<Image
						src={product?.photo?.image?.publicUrlTransformed || ''}
						alt={product?.photo?.altText || ''}
						width="80%"
						height="80%"
						loading="eager"
					/>
				) : (
					<LoadingSkeleton />
				)}
				<div style={{ display: 'flex', flex: 1, flexDirection: 'column' }}>
					<SupremeDescription>
						{product?.description || TEXT_NO_DESCRIPTION}
					</SupremeDescription>
				</div>
			</div>
		</ProductStyles>
	);
};

const ProductStyles = styled.div`
	display: grid;
	grid-auto-columns: 1fr;
	grid-auto-flow: column;
	max-width: var(--maxWidth);
	justify-content: center;
	align-items: top;
	gap: 2rem;
	img {
		width: 100%;
		object-fit: contain;
	}
`;

export default SingleProduct;
