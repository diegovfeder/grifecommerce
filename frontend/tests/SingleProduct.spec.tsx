import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { GraphQLError } from 'graphql';
import { resolveMockState } from './utils';
import { PRODUCT_QUERY } from '../gql/queries';
import SingleProduct from '../components/SingleProduct';

const mocks = [
	{
		request: {
			query: PRODUCT_QUERY,
			variables: {
				id: '718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d',
			},
		},
		result: {
			data: {
				product: {
					id: '718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d',
					name: 'Sample Pack',
					price: 2000,
					description: '',
					photo: {
						id: '0088b6c6-a3f1-4f4b-b929-3592b353d825',
						altText: 'sample_pack',
						image: {
							publicUrlTransformed:
								'https://res.cloudinary.com/grifemusic/image/upload/v1652557878/grifecommerce/cl36aaj690000gt9kejqreuiq.png',
						},
					},
				},
			},
		},
	},
];

describe('<Single Product/>', () => {
	it('renders with proper data', async () => {
		const { container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<SingleProduct id="718b7ac6-7cf1-47e7-b1de-c4a13ca92f2d" />
			</MockedProvider>,
		);
		await screen.findAllByLabelText('Loading...');
		await screen.findByTestId('loading-label');
		await resolveMockState();
		await screen.findByTestId('single-product-component');
		expect(container).toMatchSnapshot();
	});

	it('Errors out when an item is no found', async () => {
		const errorMock = [
			{
				request: {
					query: PRODUCT_QUERY,
					variables: {
						id: '123',
					},
				},
				result: {
					errors: [new GraphQLError('Product not found')],
				},
			},
		];
		const { container } = render(
			<MockedProvider mocks={errorMock} addTypename={false}>
				<SingleProduct id="123" />
			</MockedProvider>,
		);
		await screen.findByTestId('graphql-error');
		expect(container).toHaveTextContent('Error:');
		expect(container).toHaveTextContent('Product not found');
	});
});
