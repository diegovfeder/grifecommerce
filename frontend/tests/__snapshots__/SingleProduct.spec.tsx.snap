// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct/> renders with proper data 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-14jd9my-0 jPWReR"
    data-test-id="single-product-component"
  >
    <div
      style="display: flex; flex: 1; flex-direction: column;"
    >
      <h1
        style="border-bottom: 4px solid gray; font-weight: 100;"
      >
        Sample Pack
      </h1>
      <h3
        class="Supreme-sc-2anpcp-0 cenBAL"
      >
        R$ 20,00
      </h3>
      next/image stub
      <div
        style="display: flex; flex: 1; flex-direction: column;"
      >
        <h3
          class="Supreme__SupremeDescription-sc-2anpcp-1 eSreBk"
        >
          - No description available
        </h3>
      </div>
      <div
        class="buttonList"
      >
        <button
          class="StyledButtonLink-sc-zq7nji-0 chkZRh"
          type="button"
        >
          Add To Cart 🛒
        </button>
        <button
          class="StyledButtonLink-sc-zq7nji-0 chkZRh"
        >
          Update ✏️
        </button>
        <button
          class="StyledButtonLink-sc-zq7nji-0 chkZRh"
          type="button"
        >
          \`\`\`
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
